pydantic/_internal/_typing_extra.py

```diff
@@ -221,12 +221,36 @@ def eval_type_lenient(value: Any, globalns: dict[str, Any] | None, localns: dict
         value = _make_forward_ref(value, is_argument=False, is_class=True)
 
     try:
-        return typing._eval_type(value, globalns, localns)  # type: ignore
+        return eval_type_backport(value, globalns, localns)
     except NameError:
         # the point of this function is to be tolerant to this case
         return value
 
 
+def is_unsupported_types_for_union_error(e: TypeError) -> bool:
+    return str(e).startswith('unsupported operand type(s) for |: ')
+
+
+def eval_type_backport(
+    value: Any, globalns: dict[str, Any] | None = None, localns: dict[str, Any] | None = None
+) -> Any:
+    """Like `typing._eval_type`, but falls back to the `eval_type_backport` package if it's
+    installed to let older Python versions use newer typing features.
+    Currently this just means that `X | Y` is converted to `Union[X, Y]` if `X | Y` is not supported.
+    This would also be the place to add support for `list[int]` instead of `List[int]` etc.
+    """
+    try:
+        return typing._eval_type(  # type: ignore
+            value, globalns, localns
+        )
+    except TypeError as e:
+        if not (isinstance(value, typing.ForwardRef) and is_unsupported_types_for_union_error(e)):
+            raise
+        from eval_type_backport import eval_type_backport
```

Comment:
It seems to me that right now if you don't have the `eval_type_backport` package installed, you'll get a `ModuleNotFoundError` if you hit this code branch. I think that's more confusing than it needs to be; I think it would be better if you could add a `try: except ImportError:` to provide a nicer error message explaining to the user what they've done wrong (use the `|` notation for unions in older python version) and how to resolve it (use `typing.Union` or install the `eval_type_backport` package).

Reply:
Right, this was my plan.

> This PR isn't fully ready (it just assumes that the package is installed) but I want to check that this is going in the right direction.